nvme_gw_persistence.py

```diff
@@ -0,0 +1,404 @@
+#
+#  Copyright (c) 2021 International Business Machines
+#  All rights reserved.
+#
+#  SPDX-License-Identifier: LGPL-3.0-or-later
+#
+#  Authors: anita.shekar@ibm.com, sandy.kaur@ibm.com
+#
+
+import rados
+from typing import Dict, Optional
+from abc import ABC, abstractmethod
+import nvme_gw_pb2 as pb2
+
+
+class PersistentConfig(ABC):
+    """Persists gateway NVMeoF target configuration."""
+
+    @abstractmethod
+    def add_bdev(self, bdev_name: str, val: str):
+        pass
+
+    @abstractmethod
+    def delete_bdev(self, bdev_name: str):
+        pass
+
+    @abstractmethod
+    def add_namespace(self, subsystem_nqn: str, bdev_name: str, val: str):
+        pass
+
+    @abstractmethod
+    def delete_namespace(self, subsystem_nqn: str, bdev_name: str):
+        pass
+
+    @abstractmethod
+    def add_subsystem(self, subsystem_nqn: str, val: str):
+        pass
+
+    @abstractmethod
+    def delete_subsystem(self, subsystem_nqn: str):
+        pass
+
+    @abstractmethod
+    def add_host(self, subsystem_nqn: str, host_nqn: str, val: str):
+        pass
+
+    @abstractmethod
+    def delete_host(self, subsystem_nqn: str, host_nqn: str):
+        pass
+
+    @abstractmethod
+    def add_listener(self, subsystem_nqn: str, traddr: str, trsvcid: str,
+                     val: str):
+        pass
+
+    @abstractmethod
+    def delete_listener(self, subsystem_nqn: str, traddr: str, trsvcid: str):
+        pass
+
+    @abstractmethod
+    def set_allow_all(self, subsystem_nqn: str, val: str):
+        pass
+
+    @abstractmethod
+    def delete_allow_all(self, subsystem_nqn: str):
+        pass
+
+    @abstractmethod
+    def set_transport(self, trtype: str, val: str):
+        pass
+
+    @abstractmethod
+    def delete_transport(self, trtype: str):
+        pass
+
+    @abstractmethod
+    def delete_config(self):
+        pass
+
+    @abstractmethod
+    def restore(self, callbacks):
+        pass
+
+
+class OmapPersistentConfig(PersistentConfig):
+    """Persists NVMeoF target configuration to an OMAP object.
+
+    Handles reads/writes of persistent NVMeoF target configuration data in 
+    key/value format within an OMAP object.
+
+    Class attributes:
+        X_KEY: OMAP key name for "X"
+        X_PREFIX: OMAP key prefix for key of type "X"
+
+    Instance attributes:
+        version: Local gateway NVMeoF target configuration version
+        nvme_config: Basic gateway parameters
+        logger: Logger instance to track OMAP access events
+        spdk_rpc: Module methods for SPDK
+        spdk_rpc_client: Client of SPDK RPC server
+        omap_name: OMAP object name
+        ioctx: I/O context which allows OMAP access
+    """
+
+    OMAP_VERSION_KEY = "omap_version"
+    BDEV_PREFIX = "bdev_"
+    NAMESPACE_PREFIX = "namespace_"
+    SUBSYSTEM_PREFIX = "subsystem_"
+    HOST_PREFIX = "subsystem_host_"
+    ALLOW_ALL_PREFIX = "allow_all_"
```

Comment:
I am not sure "alllow_all_" is a good prefix. This is just a special case of subsystem "host" configuration. Can we use `HOST_PREFIX` here and just have a special (wildcard) name for the host, e.g. "*"?

Actually, the same question for server/cli. For hosts configuration, in cli right now we have only `allow_any_hosts`. In server we have `nvmf_subsystem_allow_any_host` and `nvmf_subsystem_add_host`, so I guess we are going tol add `add_host` to the cli too. But I don't think it is necessary to mimic the tgt commands. We could have just `add_host` command and use a special value for "any".